### PR TITLE
Refactor plugin registration and delete dumb plugins

### PR DESCRIPTION
**Pull request by hoxworth**

### CODE REVIEW

**Comment by markolson** [comment]
I HAVE NO GREEN BUTTON WITH WHICH TO MERGE THIS

**Comment by hoxworth** [comment]
Because Robb is a jerk

**Comment by hoxworth** [comment]
There green button

**Comment by robbkidd** [comment]
What did I do? 😇

Plugin specs need updating.

**Comment by robbkidd** [comment]
### First Change 

I checked this branch out and tweaked the [MockPlugin](https://github.com/markolson/linkbot/blob/mark-refactor/spec/plugins/mock.rb#L17) to use the new registration DSL.

```ruby
register :regex => //
help '!mockery - mock plugin that only repeats messages'
```

Test failure is then:

```
  1) Linkbot::Plugin has the correct plugin
     Failure/Error: expect(Linkbot::Plugin.plugins['mock'][:ptr]).to eq MockPlugin
     NoMethodError:
       undefined method `[]' for nil:NilClass
     # ./spec/plugins_spec.rb:12:in `block (2 levels) in <top (required)>'
```

```
[1] pry(#<RSpec::ExampleGroups::LinkbotPlugin>)> Linkbot::Plugin.plugins
=> {"MockPlugin"=>{:ptr=>MockPlugin, :handlers=>{:message=>{:regex=>//, :handler=>:on_message}}}}
```

### Second Change

Possibly this test is the only thing expecting to be able to look a plugin up by name. Changing `plugins['mock']` to `plugins['MockPlugin']` gets past that error and on to ...

```
  1) Linkbot::Plugin has the correct plugin
     Failure/Error: expect(Linkbot::Plugin.plugins['MockPlugin'][:handlers]).to eq({:message => {:regex => //, :handler => :on_message, :help => :help}})

       expected: {:message=>{:regex=>//, :handler=>:on_message, :help=>:help}}
            got: {:message=>{:regex=>//, :handler=>:on_message}}
```
### Third Change

Remove `, :help => :help` from the expectation and then the tests are green.

### Postscript

I find myself missing intra-repo branches for this instead of personal forks. I could just step on your toes by pushing these edits myself then. :smiley:

**Comment by robbkidd** [comment]
The patch if that makes the changes more clear:

```
diff --git a/spec/plugins/mock.rb b/spec/plugins/mock.rb
index f463f51..90d7a30 100644
--- a/spec/plugins/mock.rb
+++ b/spec/plugins/mock.rb
@@ -2,8 +2,6 @@ class MockPlugin < Linkbot::Plugin
   @@messages = []
   @@matches = []

-  def self.help; HELP; end
-
   def self.on_message(message, matches)
     @@messages << message
     @@matches << matches
@@ -14,5 +12,6 @@ class MockPlugin < Linkbot::Plugin
   def self.messages; @@messages; end
   def self.matches; @@matches; end

-  Linkbot::Plugin.register('mock', self, {:message => {:regex => //, :handler => :on_message, :help => :help}})
+  register :regex => //
+  help '!mockery - mock plugin that only repeats messages'
 end
diff --git a/spec/plugins_spec.rb b/spec/plugins_spec.rb
index ec379bd..2ddd43e 100644
--- a/spec/plugins_spec.rb
+++ b/spec/plugins_spec.rb
@@ -9,12 +9,12 @@ describe Linkbot::Plugin do

   it "has the correct plugin" do
     expect(Linkbot::Plugin.plugins.keys.length).to eq 1
-    expect(Linkbot::Plugin.plugins['mock'][:ptr]).to eq MockPlugin
-    expect(Linkbot::Plugin.plugins['mock'][:handlers]).to eq({:message => {:regex => //, :handler => :on_message, :help => :help}})
+    expect(Linkbot::Plugin.plugins['MockPlugin'][:ptr]).to eq MockPlugin
+    expect(Linkbot::Plugin.plugins['MockPlugin'][:handlers]).to eq({:message => {:regex => //, :handler => :on_message}})
   end

   it "handles a message" do
-    plugin = Linkbot::Plugin.plugins['mock'][:ptr]
+    plugin = Linkbot::Plugin.plugins['MockPlugin'][:ptr]
     response = Linkbot::Message.handle(message)

     expect(plugin.messages.length).to eq 1
```

File: plugins/no.rb [hunk 1]
```diff
@@ -1,77 +0,0 @@
-class No < Linkbot::Plugin
```
Comment (robbkidd) [inline plugins/no.rb]: HEY!

**Comment by hoxworth** [comment]
@robbkidd Thanks, fixed the plugin test. I didn't know there even WAS a plugin test! And yeah, I got rid of the whole "plugin name" thing and just made it grab the name of the class - you can still explicitly set the plugin name, but I figure that's just kinda stupid.

**Comment by robbkidd** [comment]
I dig the new plugin registration DSL. :100:

In reconsidering the plugin test, the first assertion/expectation is that there is only one plugin registered. The subsequent tests could just grab `Linkbot::Plugin.plugins.first` and maybe we don't have to deal with what plugins are named because "in production" we don't care about their names.

**Comment by robbkidd** [comment]
And reconsidering my reconsidering, we can change the test to be more generic when that test bites us again.

:+1: to merge

**Comment by robbkidd** [comment]
Living on the wild side. :shipit:

**Pull request merged**